internal/typeexpr/get_type_test.go

```diff
@@ -401,3 +424,243 @@ func TestGetTypeJSON(t *testing.T) {
+					"a": cty.String,
+					"b": cty.Number,
+				}, []string{"b"})),
+				Children: map[string]*Defaults{
+					"": {
+						Type: cty.ObjectWithOptionalAttrs(map[string]cty.Type{
+							"a": cty.String,
+							"b": cty.Number,
+						}, []string{"b"}),
+						DefaultValues: map[string]cty.Value{
+							"b": cty.NumberIntVal(5),
+						},
+					},
+				},
+			},
+			"",
+		},
+		{
+			`set(object({ a = string, b = optional(number, 5) }))`,
+			&Defaults{
+				Type: cty.Set(cty.ObjectWithOptionalAttrs(map[string]cty.Type{
+					"a": cty.String,
+					"b": cty.Number,
+				}, []string{"b"})),
+				Children: map[string]*Defaults{
+					"": {
+						Type: cty.ObjectWithOptionalAttrs(map[string]cty.Type{
+							"a": cty.String,
+							"b": cty.Number,
+						}, []string{"b"}),
+						DefaultValues: map[string]cty.Value{
+							"b": cty.NumberIntVal(5),
+						},
+					},
+				},
+			},
+			"",
+		},
+
+		// tuples containing objects with defaults work differently from
+		// collections
+		{
+			`tuple([string, bool, object({ a = string, b = optional(number, 5) })])`,
+			&Defaults{
+				Type: cty.Set(cty.ObjectWithOptionalAttrs(map[string]cty.Type{
+					"a": cty.String,
+					"b": cty.Number,
+				}, []string{"b"})),
+				Children: map[string]*Defaults{
+					"2": {
+						Type: cty.ObjectWithOptionalAttrs(map[string]cty.Type{
+							"a": cty.String,
+							"b": cty.Number,
+						}, []string{"b"}),
+						DefaultValues: map[string]cty.Value{
+							"b": cty.NumberIntVal(5),
+						},
+					},
+				},
+			},
+			"",
+		},
+
+		// incompatible default value causes an error
+		{
+			`object({ a = optional(string, "hello"), b = optional(number, true) })`,
+			&Defaults{
+				Type: cty.ObjectWithOptionalAttrs(map[string]cty.Type{
+					"a": cty.String,
+					"b": cty.Number,
+				}, []string{"a"}),
+				DefaultValues: map[string]cty.Value{
+					"a": cty.StringVal("hello"),
+				},
+			},
+			"This default value is not compatible with the attribute's type constraint: number required.",
+		},
+
+		// Too many arguments
+		{
+			`object({name=string,meta=optional(string, "hello", "world")})`,
+			nil,
+			`Optional attribute modifier expects at most two arguments: the attribute type, and a default value.`,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.Source, func(t *testing.T) {
+			expr, diags := hclsyntax.ParseExpression([]byte(test.Source), "", hcl.Pos{Line: 1, Column: 1})
+			if diags.HasErrors() {
+				t.Fatalf("failed to parse: %s", diags)
+			}
+
+			_, got, diags := getType(expr, true, true)
+			if test.WantError == "" {
+				for _, diag := range diags {
+					t.Error(diag)
+				}
+			} else {
+				found := false
+				for _, diag := range diags {
+					t.Log(diag)
+					if diag.Severity == hcl.DiagError && diag.Detail == test.WantError {
+						found = true
+					}
+				}
+				if !found {
+					t.Errorf("missing expected error detail message: %s", test.WantError)
+				}
+			}
+
+			if !cmp.Equal(test.Want, got, ignoreOpts) {
```

Comment:
This needs a `cmp.Comparer` for equality, otherwise we're just checking equality of empty structs after the unexported fields are ignored. There is already one declared in the package as `valueComparer`, eo e.g.

```
if !cmp.Equal(test.Want, got, ignoreOpts, valueComparer) {
```

Reply:
Yikes! Thanks for catching that. I had to add a `typeComparer` to ensure we actually compare the types in the test, which caught a couple of completely incorrect `test.Want` values. Fixed in the last push.